buildSrc/build.gradle.kts

```diff
@@ -60,6 +60,7 @@ dependencies {
   implementation("org.jetbrains.kotlin:kotlin-gradle-plugin:1.9.23")
   implementation("org.owasp:dependency-check-gradle:9.0.9")
   implementation("ru.vyarus:gradle-animalsniffer-plugin:1.7.1")
+  implementation("org.spdx:spdx-gradle-plugin:0.5.0")
```

Comment:
I'm worried about the overall maturity of this whole tool chain:

- Based on the [github repo](https://github.com/spdx/spdx-gradle-plugin) this plugin is quite new and seldom used
- According to the [link](https://github.com/security-slam/metrics/blob/main/cncf/2023/mechanizer.md), SBOM is supposed to be paired with a [SECURITY_INSIGHTS.yml](https://github.com/ossf/security-insights-spec/blob/main/specification.md) file, consisting of structured information on how a consumer could find / validate SBOM contents in an automated way. This too seems to be a very young technology.

Is there another java project which has implemented this automation that we can go inspect as an example? I'm really stuck on figuring out how all this stuff would actually get used. I don't really care about advertising that we've checked a box if its not usable in a practical sense. Bundling all the SBOMs into a zip in a location which is not well known would not be consumable in an automated way.

Reply:
For CNCF is you want that Mechanized badge, you need to make an SBOM and then point to it in that insights file.  This PR is just the first step - get the SBOM created.  Users have already requested sboms.  This would allow then to  use the sbom in whatever vuln tool chains they have.  Creating that insights yaml would be useful, but I'm not familiar with how much users are asking for that or how widely it's being used.

Reply:
> Users have already requested sboms.

Can you point me to this for the opentelemetry-java project? I'd like to understand how these users expect to consume the SBOM, and whether bundling all SBOMs together in a zip is common practice and something that can be integrated toolchains automatically or with low effort.

Reply:
> Based on the [github repo](https://github.com/spdx/spdx-gradle-plugin) this plugin is quite new and seldom used

This tool was first released May 2023 and has had [multiple releases](https://github.com/spdx/spdx-gradle-plugin/tags) since then. If you're concerned that it's still not 1.0, I think OTel in general is hardly one to be complaining about that. 
As for it being seldom used, from what I've seen of the various SBOM generators this seems to generate the best results for our project. If you have another option you'd rather consider please let me know.

> whether bundling all SBOMs together in a zip is common practice

I don't know if this is a common practice, but this is something we decided can easily be done for the various repos across OpenTelemetry and documented on the website.  If there is subsequent demand for publishing the SBOMs somewhere specific for Java, that can be done as a follow-on task.

This PR is an attempt to get something started in a way that can be iterated on based on feedback from users.  It's not trying to be compatible with any particular automated validation tooling.

Reply:
I wanted another example of an oss project which publishes SBOMs to compare our implementation to theirs. Did some searching and discovered that [apache arrow](https://github.com/apache/arrow/tree/main/java) does this.

For an example of how they publish, see the the `org.apache.arrow:arrow-vector:15.0.2` publication artifacts here: https://repo1.maven.org/maven2/org/apache/arrow/arrow-vector/15.0.2/

Notice the [arrow-vector-15.0.2-cyclonedx.json](https://repo1.maven.org/maven2/org/apache/arrow/arrow-vector/15.0.2/arrow-vector-15.0.2-cyclonedx.json) and corresponding `.asc`, `.md5`, `.sha1` files. It looks like they also generate an xml equivalent.

This [issue](https://cwiki.apache.org/confluence/display/COMDEV/SBOM) links to some other notable java projects which publish SBOMs, including avro, flink, hive, etc. A quick glance at their maven artifacts shows the same strategy as arrow.

Let's try to mimic this publication strategy.

Reply:
I'm willing to work on publishing these SBOM's to maven central, but I'd like to do so in a separate PR. Can this PR be merged while I start working on that?

Reply:
I'm not ready to merge this. 

This PR uses the `org.spdx.sbom` plugin, but there's also a more popular [org.cyclonedx.bom](https://github.com/CycloneDX/cyclonedx-gradle-plugin) plugin that seems to be used by other OSS projects like the ones I referenced above.

Here's an example of the output of `org.spdx.sbom`:

[opentelemetry-java_opentelemetry-api.spdx.json](https://github.com/open-telemetry/opentelemetry-java/files/14727095/opentelemetry-java_opentelemetry-api.spdx.json)

And here's an example of the output of `or.cyclonedx.bom`:

[arrow-vector-15.0.2-cyclonedx.json](https://github.com/open-telemetry/opentelemetry-java/files/14727103/arrow-vector-15.0.2-cyclonedx.json)

Its a completely different format. Based on the `sbom-format` field described in [this document describing the contents of SECURITY-INSIGHTS.yml](https://github.com/ossf/security-insights-spec/blob/main/specification.md#dependencies), there are different sbom formats, with SPDX and CycloneDX being two of them. 

We need to choose the right one because the format becomes part of our publication contract. Once users start to build tooling that depends on a format, changing it is disruptive and might be considered breaking (not sure - need to think more and discuss). I'm leaning towards CycloneDX because:

- It is more popular with more examples
- The format makes more sense to me when I look at the `.json` file. I can see how I could build automation to verify the software lifecycle based on the CycloneDx format. Not sure how I would actually use the SPDX format to do the same

Also, in order to accept the argument to merge this now and followup with more, I need a high degree of confidence that project will be seen through. I'm still waiting on followups on the [benchmarking](https://github.com/open-telemetry/opentelemetry-java/pull/5842#pullrequestreview-1645188955) project, and so I don't give the benefit of the doubt to this.

Reply:
It turned out to be easier than I thought to include the sbom with the existing maven publish.  I verified this using the `publishToMavenLocal` task.

As for the SPDX vs CycloneDX question, there was some evaluation of various plugins as shown in this PR's history and other language repos are also going with SPDX. I didn't try the gradle specific plugin for cyclonedx, so I can look into going through this same process and publish both formats.

Regarding the benchmarking question the original project goal was to establish a recurring benchmark that evaluated span generation and serialization since those are the most impactful to CPU overhead on the SDK side.  I think the existing selection I chose covered that scenario.  It wasn't our goal to have comprehensive benchmarking as automating that would likely overburden the shared runner.

Reply:
> It turned out to be easier than I thought to include the sbom with the existing maven publish. I verified this using the publishToMavenLocal task.

Perfect 👍 

> As for the SPDX vs CycloneDX question, there was some evaluation of various plugins as shown in this PR's history and other language repos are also going with SPDX

Let's chat at the next java SIG. Should be straight forward to use the CycloneDx if needed, but let's discuss with the group before doing more work.

Reply:
I went ahead and added it anyway.  Now both formats get included in the zip and published to maven central.
One problem... Inside the cyclonedx file, I wasn't able to configure the name to match what is actually published for things like `all` -> `api`.  (We have quite a few places where the published name doesn't match the folder/project name.)

Any other concerns?

Reply:
Thanks for doing that. As an example of the output, below I've included copies of the spdx and cyclonedx for two artifacts, `opentelemetry-context` which has no external dependencies, and `opentelemetry-exporter-otlp` which has several external dependencies.

`opentelemetry-context`:

[opentelemetry-context-1.37.0-SNAPSHOT-cyclonedx.json](https://github.com/open-telemetry/opentelemetry-java/files/14746262/opentelemetry-context-1.37.0-SNAPSHOT-cyclonedx.json)
[opentelemetry-context-1.37.0-SNAPSHOT-spdx.json](https://github.com/open-telemetry/opentelemetry-java/files/14746264/opentelemetry-context-1.37.0-SNAPSHOT-spdx.json)

`opentelemetry-exporter-otlp`:

[opentelemetry-exporter-otlp-1.37.0-SNAPSHOT-cyclonedx.json](https://github.com/open-telemetry/opentelemetry-java/files/14746294/opentelemetry-exporter-otlp-1.37.0-SNAPSHOT-cyclonedx.json)
[opentelemetry-exporter-otlp-1.37.0-SNAPSHOT-spdx.json](https://github.com/open-telemetry/opentelemetry-java/files/14746295/opentelemetry-exporter-otlp-1.37.0-SNAPSHOT-spdx.json)

I think including both is both conservative and aggressive in some ways. Its conservative because it ensures that users interested in either format get what they need. Its aggressive because it adds more publishing surface area than what we actually might need if we were to wait for user feedback.

On the whole, I'm in favor of publishing both, but curious of what other @open-telemetry/java-approvers think.